tokio/src/sync/broadcast.rs

```diff
@@ -897,15 +898,19 @@ impl<T> Shared<T> {
         'outer: loop {
             while wakers.can_push() {
                 match list.pop_back_locked(&mut tail) {
-                    Some(mut waiter) => {
-                        // Safety: `tail` lock is still held.
-                        let waiter = unsafe { waiter.as_mut() };
-
-                        assert!(waiter.queued);
-                        waiter.queued = false;
-
-                        if let Some(waker) = waiter.waker.take() {
-                            wakers.push(waker);
+                    Some(waiter) => {
+                        unsafe {
+                            // Safety: accessing `waker` is safe because
+                            // the tail lock is held.
+                            if let Some(waker) = (*waiter.as_ptr()).waker.take() {
+                                wakers.push(waker);
+                            }
+
+                            // Safety: `queued` is atomic.
+                            // `Release` is needed to synchronize with `Recv::drop`.
+                            // It is critical to set this variable **after** waker
+                            // is extracted, otherwise we may data race with `Recv::drop`.
+                            assert!((*waiter.as_ptr()).queued.swap(false, Release));
```

Comment:
This `swap` is equivalent to `store`.

Reply:
Do you think it would be better to use `load` for the assertion and then `store`? Or simply remove the assertion?

Reply:
Ah, I missed the assertion.

A load-then-store is fine.